File: 2.AddTwoNumbers/addtwonumbers.cc
/*
 * Definition for singly-linked list.
 *
 *
 * struct ListNode {
 *   int val;
 *   ListNode *next;
 *   ListNode(int x):val(x),next(NULL){}
 * };
 */

class Solution{
 public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2){
    ListNode *rst,*pre,*cur;
    int carry = 0,sum;

    if(l1==NULL) return l2;
    if(l2==NULL) return l1;

    // alloc a useless head node, just to simplify program
    rst = new ListNode(0);
    pre = cur = rst;
    
    while(l1 && l2){
      sum = l1->val + l2->val + carry;
      carry = sum / 10;
      sum %= 10;
      cur = new ListNode(sum);
      pre->next = cur;
      pre = cur;
      
      l1 = l1->next;
      l2 = l2->next;
    }

    while(l1){
      sum = l1->val + carry;
      carry = sum / 10;
      sum %= 10;
      cur = new ListNode(sum);
      pre->next = cur;
      pre = cur;
      l1 = l1->next;
    }

    while(l2){
      sum = l2->val + carry;
      carry = sum / 10;
      sum %= 10;
      cur = new ListNode(sum);
      pre->next = cur;
      pre = cur;
      l2 = l2->next;
    }

    if(carry){
      cur = new ListNode(carry);
      pre->next = cur;
    }

    /*del the useless head node*/
    cur = rst;
    rst = rst->next;
    delete cur;
    return rst;
  }
};

File: algorithms/src/2.AddTwoNumbers/java/AddTwoNumbers.java
/**
 * You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse
 * order and each of their nodes contain a single digit.
 * <p>
 * Add the two numbers and return it as a linked list.
 * You may assume the two numbers do not contain any leading zero, except the number 0 itself.
 * <p>
 * Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
 * Output: 7 -> 0 -> 8
 *
 * @author xingdl2007
 * @author xingdl2007@gmail.com
 */

/**
 * Definition for singly-linked list.
 *
 * @code public class ListNode {
 * @code int val;
 * @code ListNode next;
 * @code ListNode(int x) { val = x; }
 * @code }
 */
public class AddTwoNumbers {
    static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }

        public String toString() {
            return Integer.toString(val);
        }
    }

    // my accepted solution
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode(0);
        ListNode cur = head;
        int carry = 0;

        while (l1 != null && l2 != null) {
            cur.next = new ListNode((l1.val + l2.val + carry) % 10);
            carry = (l1.val + l2.val + carry) / 10;

            l1 = l1.next;
            l2 = l2.next;
            cur = cur.next;
        }

        while (l1 != null) {
            cur.next = new ListNode((l1.val + carry) % 10);
            carry = (l1.val + carry) / 10;
            l1 = l1.next;
            cur = cur.next;
        }

        while (l2 != null) {
            cur.next = new ListNode((l2.val + carry) % 10);
            carry = (l2.val + carry) / 10;
            l2 = l2.next;
            cur = cur.next;
        }

        // quiet easy to forget
        if (carry > 0) {
            cur.next = new ListNode(carry);
        }

        return head.next;
    }

    // official solution
    public ListNode addTwoNumbersCompare(ListNode l1, ListNode l2) {
        ListNode dummyHead = new ListNode(0);
        ListNode p = l1, q = l2, curr = dummyHead;
        int carry = 0;
        while (p != null || q != null) {
            int x = (p != null) ? p.val : 0;
            int y = (q != null) ? q.val : 0;
            int sum = carry + x + y;
            carry = sum / 10;
            curr.next = new ListNode(sum % 10);
            curr = curr.next;
            if (p != null) p = p.next;
            if (q != null) q = q.next;
        }
        if (carry > 0) {
            curr.next = new ListNode(carry);
        }
        return dummyHead.next;
    }

    public static void main(String[] args) {
        ListNode l1 = new ListNode(2);
        l1.next = new ListNode(4);
        l1.next.next = new ListNode(3);

        ListNode l2 = new ListNode(5);
        l2.next = new ListNode(6);
        l2.next.next = new ListNode(4);

        ListNode res = new AddTwoNumbers().addTwoNumbers(l1, l2);

        while (res != null) {
            System.out.print(res + " ");
            res = res.next;
        }
    }

}